vivarium/controllers/notebook_controller.py

```diff
@@ -513,44 +521,58 @@ def run(self, threaded=True, num_steps=math.inf):
             return
         self._is_running = True
         if threaded:
-            run_thread = threading.Thread(target=self._run, args=(num_steps,))
+            run_thread = threading.Thread(target=self._run, args=(num_steps, catch_errors))
             run_thread.daemon = True
             run_thread.start()
         else:
             self._run(num_steps=num_steps)
 
-    def _run(self, num_steps=math.inf):
+    def _run(self, num_steps=math.inf, catch_errors=True):
         """run the simulation for a given number of steps
 
         :param num_steps: num_steps, defaults to math.inf
+        :param catch_errors: wether to catch errors or not, defaults to False
         """
-        while self.time < num_steps and self._is_running:
+        # Add a local time for the run function independant from the controller time
+        run_time = 0
+        while run_time < num_steps and self._is_running:
             with self.batch_set_state():
                 # execute routines of the controller
-                self.controller_routine_step(self.time)
+                self.controller_routine_step(self.time, catch_errors=catch_errors)
 
-                # execute routines of the entities
+                # execute routines of the existing entities
                 for entity in self.all_entities:
-                    entity.routine_step(self.time)
                     entity.set_events()
+                    if not entity.exists:
+                        continue
+                    entity.routine_step(self.time, catch_errors=catch_errors)
                     # execute behaviors of agents
                     if entity.etype == EntityType.AGENT:
                         entity.behave(self.time)
+                        # TODO: see if we do this in a dedicated function or not
+                        # increment time since last meal for all alive agents
+                        entity.time_since_last_meal += 1
```

Comment:
Shouldn't this be inside the routine instead?

Reply:
In general we should avoid specific mechanisms within the generic run loop, they should all be in routines